PR for #275: acs data

gabrieleborg opened the pull request:
The PR contains all the updates to the scripts downloading ACS Census data: 
1. Added `raw/acs_zcta/code/build.py` script to fetch, combine, rename and store ACS data; 
2. ACS data is stored in `drive/raw_data/acs_rents/acs_zcta_data.csv`; 
3. Removed folder `base/acs_rents/`. Since we've automated the fetching step, we now handle table formatting in the same script (`raw/acs_zcta/code/build.py`); 
4. Related to point 3, all columns that were present in base/acs_rents/  were added maintaining the same naming format that was previously used;

Some comments on the data downloaded: 
5. Table B20026: I have not added it, but if we need there’s also the population in occupied units by time of moving (2021-2022, 2018-2020, etc). This could give us an estimate of movers, and when they moved; 
6. Table B25070: Added both total share and binned distribution for the gross rent as pct of hh income (past 12months). In the name, upper bound is excluded (e.g., `rent_share_hh_income_10_15pct` captures `[10, 14.9]`.) ; 
7. Currently, I haven't added Margin of Errors. Easy to add if we need them. 

General Consderations: 
8. At the moment I left the API key hardcoded in the script. We should consider better management practices for configs and parameters

santiagohermo commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,324 @@
+    'B25058': ['B25058_001E'],  # Median Contract Rent Quartile (Dollars) 
+    'B25059': ['B25059_001E'],  # Upper Contract Rent Quartile (Dollars) 
+    'B25060': ['B25060_001E'],  # Aggregate Contract Rent (Dollars) 
+    'B25070': ['B25070_001E', 'B25070_002E', 'B25070_003E', 
+               'B25070_004E', 'B25070_005E', 'B25070_006E', 
+               'B25070_007E', 'B25070_008E', 'B25070_009E', 
+               'B25070_010E', 'B25070_011E'],  # Gross Rent as a Percentage of Household Income in the Past 12 Months 
```
I thought these were aggregated at the census block level. Was I wrong?

santiagohermo replied:
That's great @gabrieleborg, thanks for double checking! I guess I was wrong then :)

santiagohermo commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,324 @@
+        return None
+    
+
+@sleep_and_retry
+@limits(calls=CALLS_PER_SECOND, period=1)
+def fetch_data_for_zctas(year, table, variables, offset=0):
```
Fancy...

santiagohermo commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,324 @@
+    data_path  = os.path.join(root_path, "drive", "raw_data", "acs_rents")
+    
+    # Initialize config files
+    logging.basicConfig(filename = os.path.join(output_path, 'build.log'), filemode = 'w', 
+                        format = '%(asctime)s %(message)s', level = logging.INFO, datefmt = '%Y-%m-%d %H:%M:%S')
```
I do not see the log file in the data, is it possible that it was not added?

gabrieleborg replied:
Spent one hour trying to figure out why the log file is not saved. Couldn't find a solution. @santiagohermo @diegogentilepassaro if you guys agree I'll proceed and merge to mainline anyway leaving a note in the script. After that I'll submit a PR for #274

santiagohermo commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,324 @@
+    index_cols = ['zcta', 'Year']
+    return df[index_cols + [x for x in df.columns if x not in index_cols]]
+
+
+# test_data_availability()  # Uncomment and Run if needed
+
```
Duplicated (the exact same commented code is under the main function.

```suggestion

```

gabrieleborg replied:
Good catch

santiagohermo commented on raw/acs_zcta/smsa_fips.txt:
Why was this file moved? It looks like an error.

gabrieleborg replied:
Weird I see it in there.

santiagohermo replied:
Can you bring it back to `raw/ahs`, its original location?

gabrieleborg replied:
I didn't see any `raw/ahs/` folder. So I created it and moved the `smsa_fips.txt` file in it.

santiagohermo reviewed (commented):
Thanks @gabrieleborg! This is great :)

I left some questions below.

On your general points:
- 1-4) All makes sense! Can we store the downloaded data in `raw_data/acs_zcta`? Btw, [`raw_data/acs_rents`](https://drive.google.com/drive/u/0/folders/1R54Ap0Ea1tFwPt3ocnNb7nEGiAKj8bO3) seems to have the data I downloaded manually, so we can remove it
- On 5-7. Your decisions make sense to me!
- I commented in-line on 8.

gabrieleborg commented on base/acs_rents/output/data_file_manifest.log:
```diff
@@ -1,18 +0,0 @@
-
-========================================================================== 
-File: ../../../drive/base_large/acs_rents/zcta_year.csv 
```
dropped it

gabrieleborg commented on .gitignore:
```diff
@@ -42,3 +42,5 @@ paper_slides/slides/long_slides/min_wage_rent.log
 paper_slides/slides/short_slides/state-*
 paper_slides/staged_output/output.log
+
+.vscode/
```
line added

gabrieleborg commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,324 @@
+
+
+# Your API key
+API_KEY = 'f0aadefd118cba2a1047e3f4be068bb384908cb2'
```
Yeah, we can keep it here for the moment :) Ideally yes, you'd keep it in a file that is ignored but let's focus on more important things

santiagohermo commented:
Thanks @gabrieleborg!

- It looks like there is not commit from you, which means [this](https://github.com/diegogentilepassaro/min_wage_rent/pull/286#discussion_r1696917660), [this](https://github.com/diegogentilepassaro/min_wage_rent/pull/286#discussion_r1696928249), and [this](https://github.com/diegogentilepassaro/min_wage_rent/pull/286#discussion_r1696929556) were not changed. Maybe you forgot to push them?
- On [this](https://github.com/diegogentilepassaro/min_wage_rent/pull/286#discussion_r1700173815): it's fine if there is no log file. I would like to double check the data you downloaded before merging though. As I noticed in the [previous comment](https://github.com/diegogentilepassaro/min_wage_rent/pull/286#pullrequestreview-2207613200), the downloaded data are not online. Could you:
  - Add the downloaded data in `GDrive/raw_data/acs_zcta`, and change the code so that this is the output folder?
  - Remove from the GDrive the raw data on `acs_rents` I downloaded previously, which is [here](https://drive.google.com/drive/u/0/folders/1R54Ap0Ea1tFwPt3ocnNb7nEGiAKj8bO3)

gabrieleborg commented:
That's weird because I did it and I see only the new folder and new data....
<img width="1200" alt="Screenshot 2024-08-02 at 8 45 07 AM" src="https://github.com/user-attachments/assets/7dba5b2e-2264-4550-98b2-4107534e7d47">

I'll double check what's going on.

gabrieleborg commented:
OK found out what was happening @santiagohermo...for some reason there were 2 alternative versions of the GDrive, and I was pointing to the other one. So I was the only one seeing all I was doing 😂. I have removed my symbolic link to the "alternative reality GDrive", and re-added it pointing to the one you also see. I have then re-run the script, created the ACS data - which now you should see. Please review and let me know. AND, for some reason now I there's the log!!! I have also moved the `smsa_fips.txt` to `raw/ahs/`. Let me know about that too. Thanks!

santiagohermo commented on raw/acs_zcta/code/build.py:
```diff
@@ -0,0 +1,318 @@
+        'B17001_002E': 'Persons_Below_Poverty_Level',
+        'B00001_001E': 'Unweighted_Sample_Count_Population',
+        'B00002_001E': 'Unweighted_Sample_Count_Housing',
+        'B19001_002E': 'Income_Less_10000',
+        'B19001_003E': 'Income_10000_14999',
+        'B19001_004E': 'Income_15000_19999',
+        'B19001_005E': 'Income_20000_24999',
+        'B19001_006E': 'Income_25000_29999',
+        'B19001_007E': 'Income_30000_34999',
+        'B19001_008E': 'Income_35000_39999',
+        'B19001_009E': 'Income_40000_44999',
+        'B19001_010E': 'Income_45000_49999',
+        'B19001_011E': 'Income_50000_59999',
+        'B19001_012E': 'Income_60000_74999',
+        'B19001_013E': 'Income_75000_99999',
+        'B19001_014E': 'Income_100000_124999',
+        'B19001_015E': 'Income_125000_149999',
+        'B19001_016E': 'Income_150000_199999',
+        'B19001_017E': 'Income_200000_or_more',
```
This is household income, right? My only comment is that it would be useful to call them "hh_income_.." or something, so that it's clearer. But I don't want to make you run everything again!

gabrieleborg replied:
done

santiagohermo commented on raw/acs_zcta/output/build.log:
```diff
@@ -0,0 +1,977 @@
+2024-08-02 09:07:41 Error fetching data for year 2020, table B00001, geo state:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2020/acs/acs5?get=NAME,B00001_001E&for=state:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Failed to fetch data for year 2020, table B00001 at all geographic levels
+2024-08-02 09:07:41 No data available for year 2020, table B00001
+2024-08-02 09:07:41 Attempting to fetch data for year 2020, table B00002, geo zip code tabulation area:*, offset 0
+2024-08-02 09:07:41 Error fetching data for year 2019, table B00001, geo zip code tabulation area:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2019/acs/acs5?get=NAME,B00001_001E&for=zip%20code%20tabulation%20area:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Attempting to fetch data for year 2019, table B00001, geo county:*, offset 0
+2024-08-02 09:07:41 Successfully fetched 33120 rows for year 2017, table B00001, geo zip code tabulation area:*, offset 0
+2024-08-02 09:07:41 Final ['B00001_001E'] for 2017:Table B00001 Index(['NAME', 'B00001_001E', 'state', 'zip code tabulation area'], dtype='object')
+2024-08-02 09:07:41 Attempting to fetch data for year 2017, table B00002, geo zip code tabulation area:*, offset 0
+2024-08-02 09:07:41 Successfully fetched 33120 rows for year 2016, table B00001, geo zip code tabulation area:*, offset 0
+2024-08-02 09:07:41 Final ['B00001_001E'] for 2016:Table B00001 Index(['NAME', 'B00001_001E', 'state', 'zip code tabulation area'], dtype='object')
+2024-08-02 09:07:41 Attempting to fetch data for year 2016, table B00002, geo zip code tabulation area:*, offset 0
+2024-08-02 09:07:41 Error fetching data for year 2020, table B00002, geo zip code tabulation area:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2020/acs/acs5?get=NAME,B00002_001E&for=zip%20code%20tabulation%20area:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Attempting to fetch data for year 2020, table B00002, geo county:*, offset 0
+2024-08-02 09:07:41 Error fetching data for year 2019, table B00001, geo county:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2019/acs/acs5?get=NAME,B00001_001E&for=county:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Attempting to fetch data for year 2019, table B00001, geo state:*, offset 0
+2024-08-02 09:07:41 Error fetching data for year 2020, table B00002, geo county:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2020/acs/acs5?get=NAME,B00002_001E&for=county:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Attempting to fetch data for year 2020, table B00002, geo state:*, offset 0
+2024-08-02 09:07:41 Error fetching data for year 2019, table B00001, geo state:*, offset 0: 400 Client Error:  for url: https://api.census.gov/data/2019/acs/acs5?get=NAME,B00001_001E&for=state:*&key=f0aadefd118cba2a1047e3f4be068bb384908cb2&offset=0
+2024-08-02 09:07:41 Failed to fetch data for year 2019, table B00001 at all geographic levels
```
There were some errors here. Is it something to be concerned about?

gabrieleborg replied:
Thanks for pointing that out. I double checked and I found out that (unfortunately for us) the zcta-level data on the unweighted sample count of the population and housing ends at 2018. After that there are only tables at the county or places level. So we have the years 2019-2022 missing for the `Unweighted_Sample_Count_Population` and `Unweighted_Sample_Count_Housing` variables.

santiagohermo commented on raw/acs_zcta/output/build.log:
```diff
@@ -0,0 +1,977 @@
+2024-08-02 09:08:54 Columns in the processed dataset: ['zcta', 'Year', 'NAME', 'Total_Population', 'state', 'Total_Housing_Units', 'Median_Household_Income', 'Aggregate_Household_Income', 'Persons_Below_Poverty_Level', 'Unweighted_Sample_Count_Population', 'Unweighted_Sample_Count_Housing', 'Income_Less_10000', 'Income_10000_14999', 'Income_15000_19999', 'Income_20000_24999', 'Income_25000_29999' ...
+2024-08-02 09:08:54 Number of rows in the final dataset: 398748
+2024-08-02 09:08:54 Sample of the final dataset:
+    zcta  Year         NAME  Total_Population state  Total_Housing_Units  \
+0  01832  2013  ZCTA5 01832             22121    25                 7016   
+1  01833  2013  ZCTA5 01833              8295    25                16558   
+2  01834  2013  ZCTA5 01834              6675    25                23738   
+3  01835  2013  ZCTA5 01835             13527    25                 2502   
```
If you decide to run it again I think we can delete the NAME column.

gabrieleborg replied:
done

santiagohermo commented on raw/acs_zcta/output/build.log:
```diff
@@ -0,0 +1,977 @@
+2024-08-02 09:08:54 Columns in the processed dataset: ['zcta', 'Year', 'NAME', 'Total_Population', 'state', 'Total_Housing_Units', 'Median_Household_Income', 'Aggregate_Household_Income', 'Persons_Below_Poverty_Level', 'Unweighted_Sample_Count_Population', 'Unweighted_Sample_Count_Housing', 'Income_Less_10000', 'Income_10000_14999', 'Income_15000_19999', 'Income_20000_24999', 'Income_25000_29999' ...
+2024-08-02 09:08:54 Number of rows in the final dataset: 398748
+2024-08-02 09:08:54 Sample of the final dataset:
+    zcta  Year         NAME  Total_Population state  Total_Housing_Units  \
```
If you decide to run again I wonder whether we can save a simple data manifest. Like just compute the average and std error of each variable and save it in a log file

gabrieleborg replied:
added a function to create a file manifest (didn't use gslab since I couldn't find file manifest function to use directly with python, as opposed to Stata, R, etc.)

santiagohermo reviewed (approved):
Thanks @gabrieleborg! Yeah, the old GDrive is still active. I think my @brown account will die this month, and probably the GDrive with it. Yes! Delete old references to the old one please.

Al looks good with the branch now, so I'm approving. I left a couple of comments that I think require a new run, but I don't feel too strongly about it if you think otherwise. Feel free to S&M afterwards!

gabrieleborg commented:
@santiagohermo  I answered all your questions/requests and merged the PR!
